packages/ra-ui-materialui/src/field/ArrayField.tsx

```diff
@@ -88,78 +117,56 @@ const initialState = {
  *     )
  *     TagsField.defaultProps = { addLabel: true };
  */
-export class ArrayField extends Component<
-    FieldProps & InjectedFieldProps,
-    State
-> {
-    constructor(props: FieldProps & InjectedFieldProps) {
-        super(props);
-        this.state = props.record
-            ? this.getDataAndIds(props.record, props.source, props.fieldKey)
-            : initialState;
-    }
-
-    componentWillReceiveProps(
-        nextProps: FieldProps & InjectedFieldProps,
-        prevProps: FieldProps & InjectedFieldProps
-    ) {
-        if (nextProps.record !== prevProps.record) {
-            this.setState(
-                this.getDataAndIds(
-                    nextProps.record,
-                    nextProps.source,
-                    nextProps.fieldKey
-                )
-            );
-        }
-    }
+export const ArrayField: FunctionComponent<
+    FieldProps & InjectedFieldProps & State
+> = ({
+    addLabel,
+    basePath,
+    children,
+    record,
+    sortable,
+    source,
+    fieldKey,
+    ...rest
+}) => {
+    const [ids, setIds] = useState(() => {
+        const list = get(record, source);
+        if (!list) return [];
+        return fieldKey
+            ? list.map(item => item[fieldKey])
+            : list.map(JSON.stringify);
+    });
 
-    getDataAndIds(record: object, source: string, fieldKey: string) {
+    const [data, setData] = useState(() => {
         const list = get(record, source);
-        if (!list) {
-            return initialState;
-        }
+        if (!list) return {};
         return fieldKey
-            ? {
-                  data: list.reduce((prev, item) => {
-                      prev[item[fieldKey]] = item;
-                      return prev;
-                  }, {}),
-                  ids: list.map(item => item[fieldKey]),
-              }
-            : {
-                  data: list.reduce((prev, item) => {
-                      prev[JSON.stringify(item)] = item;
-                      return prev;
-                  }, {}),
-                  ids: list.map(JSON.stringify),
-              };
-    }
+            ? list.reduce((prev, item) => {
+                  prev[item[fieldKey]] = item;
+                  return prev;
+              }, {})
+            : list.reduce((prev, item) => {
+                  prev[JSON.stringify(item)] = item;
+                  return prev;
+              }, {});
+    });
 
-    render() {
-        const {
-            addLabel,
-            basePath,
-            children,
-            record,
-            sortable,
-            source,
-            fieldKey,
-            ...rest
-        } = this.props;
-        const { ids, data } = this.state;
+    useEffect(() => {
+        const { ids, data } = getDataAndIds(record, source, fieldKey);
+        setIds(ids);
+        setData(data);
+    }, [record, source, fieldKey]);
```

Comment:
This will be slower than the initial version, as the effect will run at least once - so the getDataAndId logic will be executed at least twice. 

I see two options: 

1. add a ref for the record, to execute the effect only if the ref has changed
2. Remove the initial value in the `useState()` calls, so that getDataAndId gets only executed once - in the effect

Reply:
Done!